src/js/App/GlobalFilter/TagsModal.tsx

```diff
@@ -7,126 +7,132 @@ import debounce from 'lodash/debounce';
+  onApplyTags: (tags: CommonSelectedTag[], sids: CommonSelectedTag[]) => void;
 };
 
-const useMetaSelector = (key: 'tags' | 'workloads' | 'sid') =>
+export type IDMapper = (tag: CommonTag) => string;
+export type CellsMapper = (tag: CommonTag) => (string | number | boolean | undefined)[];
+export type DebounceCallback = (filters?: TagFilterOptions, pagination?: TagPagination) => Action;
+
+export const useMetaSelector = (key: 'tags' | 'workloads' | 'sid') =>
   useSelector<ReduxState, [boolean | unknown, number, number, number]>(({ globalFilter }) => {
     const selected = globalFilter[key];
     return [selected?.isLoaded, selected?.total || 0, selected?.page || 1, selected?.perPage || 10];
   }, shallowEqual);
 
-const TagsModal = ({ isOpen, filterTagsBy, onApplyTags, toggleModal, selectedTags }: TagsModalProps) => {
-  const intl = useIntl();
-  const [tagsSelected, setTagsSelected] = useState<GlobalFilterTag[]>([]);
-  const [sidsSelected, setSidsSelected] = useState<SID[]>([]);
-  const [filterBy, setFilterBy] = useState('');
-  const [filterSIDsBy, setFilterSIDsBy] = useState('');
-  const dispatch = useDispatch();
-  const [tagsLoaded, tagsCount, tagsPage, tagsPerPage] = useMetaSelector('tags');
-  const [sidLoaded, sidCount, sidPage, sidPerPage] = useMetaSelector('sid');
-  const tags = useSelector<ReduxState, GlobalFilterTag[]>(({ globalFilter: { tags } }) => tags?.items || []);
-  const sids = useSelector<ReduxState, SID[]>(({ globalFilter: { sid } }) => sid?.items || []);
-  const filterScope = useSelector<ReduxState, 'insights' | undefined>(({ globalFilter: { scope } }) => scope || undefined);
-  const debounceGetTags = useCallback(
-    debounce((search?: string) => {
-      dispatch(
-        fetchAllTags(
-          {
-            registeredWith: filterScope,
-            activeTags: selectedTags,
-            search,
-          },
-          { page: tagsPage, perPage: tagsPerPage }
-        )
+const usePagination = (loaded: boolean | unknown, perPage?: number, page?: number, count?: number) => {
+  return useMemo(() => {
+    if (loaded) {
+      return {
+        perPage: perPage,
+        page: page,
+        count: count,
+      };
+    }
+    return {};
+  }, [loaded, perPage, page, count]);
+};
+
+const useRow = (
+  resource: SID[] | GlobalFilterTag[],
+  loaded: boolean | unknown,
+  idMapper: IDMapper,
+  cellsMapper: CellsMapper,
+  selected?: GlobalFilterTag[]
+) => {
+  return useMemo(() => {
+    if (loaded) {
+      return flatMap(resource, ({ tags }) =>
+        tags?.map(({ tag }) => ({
+          id: idMapper(tag),
+          namespace: tag.namespace,
+          key: tag.key,
+          value: tag.value,
+          selected: selected?.find?.(({ id }) => id === idMapper(tag)),
+          cells: cellsMapper(tag),
+        }))
       );
-    }, 800),
-    []
-  );
-  const debounceGetSIDs = useCallback(
+    }
+    return [];
+  }, [resource, selected]);
```

Comment:
the `loaded` variable is missing in the dependencies

Reply:
Good point!